src/SIL.XForge.Scripture/Services/ParatextService.cs

```diff
@@ -1177,13 +1189,7 @@ public string GetNotes(UserSecret userSecret, string projectId, int bookNum)
                 return null;
 
             CommentManager manager = CommentManager.Get(scrText);
-            var threads = manager.FindThreads(
-                (commentThread) =>
-                {
-                    return commentThread.VerseRef.BookNum == bookNum;
-                },
-                true
-            );
+            var threads = manager.FindThreads(commentThread => commentThread.VerseRef.BookNum == bookNum, true);
```

Comment:
How inefficient do you think this is? It looks like we filter over all threads for every book, so we have number_of_books * number_of_threads operations. If we assume 77 books (not uncommon, I think) and 100,000 threads (probably uncommon), we get 7,700,000 operations. That's inefficient, but these comparisons should be blazing fast, so possibly there is no real issue.

However, a) I'm not super confident it doesn't affect the speed, and b) I'm noticing that we're calling `CommentManager.FindThreads`. Given the one-to-many relationship between threads and comments, I'm wondering if this is an expensive operation to be repeating. Also, we're reading properties like `VerseRef.BookNum`. Since notes can be moved after they're created, I'm guessing it would take some computation to determine what the current verse ref is (even if they're rarely moved, it would still mean you have to read and apply the operations that have happened to the note, not just read a static property). There are probably other properties that could be similarly slow to compute. Of course, I don't know this for sure; it's possible all these operations happen once in the `CommentManager` and the only slowdown here is in repeated filtering, which may be negligible.

Reply:
My observations of the code in ParatextData:

 * The lambda function is executed in a `Where()` clause in `CommentManager.FindThreads`. The code in `CommentManager.FindThreads` could be optimized - an in-memory collection is populated from the in-memory `CommentThread` cache then the clause is executed, which is not the most efficient way to do it.
 * The `VerseRef` is retrieved from the first comment in the thread, then the collection of `Comment` objects in the thread is checked from last to first for a reattached location (if there is no reattached location, every `Comment` will be checked).

This is not the most efficient way, but I think we are dealing in tens of milliseconds of inefficiency, not hundreds of milliseconds. I think the best approach would be to create the test project with 100,000 notes, and see how things work out. If the inefficiency does prove detrimental, our only option I can think of would be to implement our own `CommentManager`.

The biggest slowdowns for code are usually IO related, which is what this PR has addressed.